certora/specs/ERC20.spec

```diff
@@ -0,0 +1,252 @@
+│ Invariant: balance of address(0) is 0                                                                               │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+invariant zeroAddressNoBalance()
+    balanceOf(0) == 0;
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: only the token holder (or a permit) can increase allowance. The spender can decrease it by using it          │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule onlyHolderOfSpenderCanChangeAllowance(env e, method f){
+    requireInvariant totalSupplyIsSumOfBalances();
+
+    calldataarg args;
+    address holder;
+    address spender;
+
+    uint256 allowanceBefore = allowance(holder, spender);
+    f(e, args);
+    uint256 allowanceAfter = allowance(holder, spender);
+
+    assert (
+        allowanceAfter > allowanceBefore
+    ) => (
+        (f.selector == sig:approve(address,uint256).selector           && e.msg.sender == holder) ||
+        (f.selector == sig:permit(address,address,uint256,uint256,uint8,bytes32,bytes32).selector)
+    );
+
+    assert (
+        allowanceAfter < allowanceBefore
+    ) => (
+        (f.selector == sig:transferFrom(address,address,uint256).selector && e.msg.sender == spender) ||
+        (f.selector == sig:approve(address,uint256).selector              && e.msg.sender == holder ) ||
+        (f.selector == sig:permit(address,address,uint256,uint256,uint8,bytes32,bytes32).selector)
+    );
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: transfer behavior and side effects                                                                            │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule transfer(env e) {
+    requireInvariant totalSupplyIsSumOfBalances();
+    require nonpayable(e);
+
+    address holder = e.msg.sender;
+    address recipient;
+    address other;
+    uint256 amount;
+
+    // cache state
+    uint256 holderBalanceBefore    = balanceOf(holder);
+    uint256 holderVotingPowerBefore = delegatedVotingPower(delegatee(holder));
+    uint256 recipientBalanceBefore = balanceOf(recipient);
+    uint256 recipientVotingPowerBefore = delegatedVotingPower(delegatee(recipient));
+    uint256 otherBalanceBefore     = balanceOf(other);
+
+    // run transaction
+    transfer@withrevert(e, recipient, amount);
+
+    // check outcome
+    if (lastReverted) {
+        assert holder == 0 || recipient == 0 || amount > holderBalanceBefore ||
+            // Handle overflows in delegation, should not be possible.
+            recipientVotingPowerBefore + amount > max_uint256 || holderVotingPowerBefore < amount ;
```

Comment:
We could try to prove those, or get as close as possible. In the ideal case we don't have to even mention it because it's proved elsewhere and we use the proof. For example, the fact that you can't have an underflow [here](https://github.com/morpho-org/morpho-token-upgradeable/blob/90b1de31fad94e243a511eb33a2fc383d93a1852/src/DelegationToken.sol#L148) could be proved. The statement would be: if A delegates to B, and B != address(0), then votingPower(B) >= balanceOf(A)

Reply:
Good idea, is this blocking the PR? Or could we handle this in a followup issue/PR?